clients/drcachesim/tracer/raw2trace.h

```diff
@@ -1073,33 +1080,66 @@ template <typename T> class trace_converter_t {
                 (instr->reads_memory() || instr->writes_memory()) &&
                 // No following memref for instruction-only trace type.
                 !is_instr_only_trace) {
-                for (uint j = 0; j < instr->num_mem_srcs(); j++) {
-                    error = append_memref(tls, &buf, instr, instr->mem_src_at(j), false,
-                                          reg_vals);
-                    if (!error.empty())
-                        return error;
-                    error = handle_kernel_interrupt_and_markers(
-                        tls, &buf, cur_modoffs, instr->length(), instrs_are_separate,
-                        &interrupted);
-                    if (!error.empty())
-                        return error;
-                    if (interrupted)
-                        break;
-                }
-                // We break before subsequent memrefs on an interrupt, though with
-                // today's tracer that will never happen (i#3958).
-                for (uint j = 0; !interrupted && j < instr->num_mem_dests(); j++) {
-                    error = append_memref(tls, &buf, instr, instr->mem_dest_at(j), true,
-                                          reg_vals);
-                    if (!error.empty())
-                        return error;
-                    error = handle_kernel_interrupt_and_markers(
-                        tls, &buf, cur_modoffs, instr->length(), instrs_are_separate,
-                        &interrupted);
-                    if (!error.empty())
-                        return error;
-                    if (interrupted)
-                        break;
+                if (instr->is_scatter_or_gather()) {
+                    // The instr should either load or store, but not both. Also,
+                    // it should have a single src or dest operand.
+                    DR_ASSERT(instr->num_mem_srcs() + instr->num_mem_dests() == 1);
+                    bool is_scatter = instr->num_mem_dests() == 1;
+                    bool reached_end_of_memrefs = false;
+                    // For expanded scatter/gather instrs, we do not have prior knowledge
+                    // of the number of store/load memrefs that will be present. So we
+                    // continue reading entries until we find a non-memref entry.
+                    while (!reached_end_of_memrefs) {
+                        // XXX: Add sanity check for max count of store/load memrefs
+                        // possible for a given scatter/gather instr.
+                        if (is_scatter) {
+                            error =
+                                append_memref(tls, &buf, instr, instr->mem_dest_at(0),
```

Comment:
This is passing the same memref_summary_t for each one.  Are we sure that won't break something?    What if we add further fields in the future: would that ever trip over this corner case?  Worth comments here and in memref_summary_t.

Reply:
I'm worried about interactions with identify_elidable_addresses.  If the tracer is looking at the expanded scatter memrefs and decides to elide something, raw2trace has to match that.

Reply:
> This is passing the same memref_summary_t for each one. Are we sure that won't break something? 

My rationale was that: if we had a separate pc entry preceding each memref, the `memref_summary_t` object for each memref would still be identical because it's based on the same instr at the same pc entry. So passing the same `memref_summary_t`  object should be equivalent.

> Worth comments here and in memref_summary_t.

Added.

Reply:
> I'm worried about interactions with identify_elidable_addresses

Good point. Can we just disable elision for scatter gather bbs?

Reply:
> My rationale was that: if we had a separate pc entry preceding each memref, the memref_summary_t object for each memref would still be identical because it's based on the same instr at the same pc entry. So passing the same memref_summary_t object should be equivalent.

It's based on the same instr right now in raw2trace b/c it's not expanding the scatter/gather.  But the instrumentation is not based on a single instr: it's based on a series of separate loads/stores.  And the elision is as well.  So we have a discrepancy and potential for problems.

> > I'm worried about interactions with identify_elidable_addresses
>
> Good point. Can we just disable elision for scatter gather bbs?

I think we should disable for now.  A possible solution is to call drx_expand_scatter_gather in raw2trace (making sure it can run statically); I would disable with an XXX comment about maybe doing this.  Are the addresses possibly reconstructable or are the index values too runtime-dependent?

Reply:
Disabled address elision for scatter/gather bbs.

Reply:
`vpscatterdd [arr + xmm11*4]{k1}, xmm10`

The memrefs do share `arr`, but the index values in `xmm11` could be anything.